### PR TITLE
Add get statistics and get charging profile

### PR DESCRIPTION
**Pull request by robthebold**
Added functions to get vehicle's statistics and the "weekly planner" or charging profile.
Some extra classes and lists still need to be added to vehicle.py to facilitate iterating through some of the results of these functions to make them more usable.
Lasttrip and Alltrips have some useful data about driving style and energy use that may only apply to electric, hybrid and plug-in hybrid vehicles.
Destinations returns all the navigation destinations stored in the vehicle nav system.
Charging profile (aka weekly planner) returns the 3 preset charging/precondition timers + the override (one-time) timer. These only apply to electric and PHEVs
Rangemap returns a set of lat/lon pairs (and center point) that define the driving range of an EV. These can be plugged into a polyline mapper (e.g. mapbox) to show a graphic representation of driving range.

### CODE REVIEW

**Comment by robthebold** [comment]
It looks like we should also be able to modify and send a revised charging schedule to the car (3 "weekly planners" + the one-off "set departure time"). But we'll need a more general command structure than what's currently in remote_services.py -- right now the only data that can be sent with the request is the name of the service to be called.

File: bimmer_connected/vehicle.py [hunk 1]
```diff
@@ -208,6 +208,61 @@ def get_vehicle_image(self, width: int, height: int, direction: VehicleViewDirec
         response = self._account.send_request(url, headers=header)
         return response.content
 
+    def get_vehicle_lasttrip(self) -> json:
```
[the remark this answers is not in the document]
Comment (robthebold) [reply bimmer_connected/vehicle.py]: My intention was to return a json. Am I doing it wrong? I was intending to return a part of the response as json, but had I converted it to dict? Python's strong-yet-weak typing is still new to me.

[the remark this answers is not in the document]
Comment (rikroe) [reply bimmer_connected/vehicle.py]: Don't have your code running atm, but generally, if you're using python's print or logging function and get a JSON-like object with single quotes `{'k': 'v'}` it's a python `dict` object. Same thing with double quotes (and of type `str`) is the JSON notation (`{"k": "v"}`, i.e. what you get using `json.dumps()` in your code in `cli.py`). To be 100% sure, either print/log `type(your_returned_variable)` which gives either `dict` or `str`.

This separation (dict in the vehicle module and JSON output on cli) is perfectly fine for me but I would strongly advise against having JSON returns in the actual model - you'd have to parse it again if you want to use it in further python code.

I think everything is correct except the return type of your functions which should be declared as dict :)

[the remark this answers is not in the document]
Comment (robthebold) [reply bimmer_connected/vehicle.py]: Finally got chance to make changes and test for breakage. Return types of python dicts are now dicts and test out in cli.py

**Comment by Markinus** [comment]
Hi,
I like this static infos and I would like to use it. Is someone still working on it?
There is a change what I would suggest. These new methods in the verhicle.py are maybe not needed. Perhaps it would better to move these statistic data receiving calls from these functions to update_data in state.py. I think this would be the best place to catching all vehicles data in one step. Then it would be possible to add some new properties in this file too.
I could do some work here if needed...I have a very new car and I can test much of this stuff.
Btw. I get an error from get_vehicle_rangemap() (only from this one): UNKNOWN_ERROR (received status code 424 and expected 200). Does someone get it too?

**Comment by robthebold** [comment]
I think bmw changed the format of the rangemap request around the beginning of the year.which broke the function. I _suspect_ that it now requires specifying which map points (comfort, ecopro) to receive instead of just returning all available.

**Comment by Markinus** [comment]
I found the problem, the error message is:
 **(SmartPhoneUtil-A-2015) Dependency not met. Vehicle tracking is disabled, range map cannot be calculated!**
And in the attributes of my car:
"rangeMap": "NOT_SUPPORTED"

So it would be possible to check this. For my car "all trips" and "charging profiles" are working. Last trip and destinations are empty.

**Comment by Markinus** [comment]
@robthebold Do you maybe have some json outputs of lastTrip, rangeMap or Destinations?

**Comment by robthebold** [comment]
> @robthebold Do you maybe have some json outputs of lastTrip, rangeMap or Destinations?

I'm away from my development machine (and my i3) for a few days, but the bmw-i-remote project on github has sample json responses - there's a link in this project's readme, that should be helpful.

**Comment by Markinus** [comment]
Ok, thanks. I will prepare a alternative kind of implementation and a PR.

**Comment by Markinus** [comment]
New PR is prepared:
https://github.com/bimmerconnected/bimmer_connected/pull/191

**Comment by gerard33** [comment]
Closing this PR as #191 is merged.

**Pull request closed without merging**